packages/flutter_tools/bin/xcode_backend.dart

```diff
@@ -326,6 +350,92 @@ class Context {
     }
   }
 
+  /// Returns `true` if a directory exists at `FLUTTER_FRAMEWORK_SWIFT_PACKAGE_PATH`.
+  bool _usingFlutterFrameworkSwiftPackage() {
+    final String? swiftPackagePath = environment['FLUTTER_FRAMEWORK_SWIFT_PACKAGE_PATH'];
+    if (swiftPackagePath == null) {
+      return false;
+    }
+    final swiftPackage = Directory(swiftPackagePath);
+    return swiftPackage.existsSync();
+  }
+
+  /// Returns `true` if the Flutter/FlutterMacOS framework Info.plist in [builtProductsDir] and
+  /// [targetBuildDir] matches the framework Info.plist from the engine cache for the
+  /// corresponding [platform] and [buildMode].
+  ///
+  /// The Info.plist contains the build mode and engine version.
+  ///
+  /// This validation will always fail when using a local engine since the engine version will not match.
+  bool _validateFlutterFramework({
+    required String buildMode,
+    required TargetPlatform platform,
+    required String? builtProductsDir,
+    required String? targetBuildDir,
+  }) {
+    if (builtProductsDir == null || targetBuildDir == null) {
+      return false;
+    }
+    try {
+      final infoPlistFromBuild = File('$builtProductsDir/${platform.infoPlistPath}');
+      if (!infoPlistFromBuild.existsSync()) {
+        return false;
+      }
+      final infoPlistFromEmbedded = File('$targetBuildDir/${platform.infoPlistPath}');
+      if (!infoPlistFromEmbedded.existsSync()) {
+        return false;
+      }
+
+      final File? infoPlistFromEngineCache = _infoPlistFromEngineCache(buildMode, platform);
+      if (infoPlistFromEngineCache == null || !infoPlistFromEngineCache.existsSync()) {
+        return false;
+      }
+      final String expectedInfoPlist = infoPlistFromEngineCache.readAsStringSync();
+      return infoPlistFromBuild.readAsStringSync() == expectedInfoPlist &&
+          infoPlistFromEmbedded.readAsStringSync() == expectedInfoPlist;
+    } on Exception catch (e) {
+      // Use `echo` instead of `echoError` so it does not cause the build to fail.
+      echo('An error occured while validating the Flutter framework: $e');
+    }
+    return false;
+  }
+
+  /// Find the Info.plist of the Flutter/FlutterMacOS framework for the corresponding [buildMode]
+  /// and [platform].
+  File? _infoPlistFromEngineCache(String buildMode, TargetPlatform platform) {
+    final String artifactMode = buildMode == 'debug'
+        ? platform.artifactName
+        : '${platform.artifactName}-$buildMode';
+    final xcframework = Directory(
+      '${environment['FLUTTER_ROOT'] ?? ''}/bin/cache/artifacts/engine/$artifactMode/${platform.frameworkName}.xcframework',
+    );
+    switch (platform) {
+      case TargetPlatform.ios:
+        final String? sdkRoot = environment['SDKROOT']?.toLowerCase();
+        if (sdkRoot == null || !sdkRoot.contains('iphone')) {
+          return null;
+        }
+        final bool simulatorSDK = sdkRoot.contains('simulator');
```

Comment:
This is mimicking:

https://github.com/flutter/flutter/blob/f63cf73d310065a11b67d82ba88618e342227c7c/packages/flutter_tools/lib/src/macos/xcode.dart#L285-L293